### ISSUE TITLE
decide on a postgres version to test against

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#10681]
Postgres changes some stuff in their protocol between versions, and also the OIDs of things between versions. Examples:
- the `verify_function_sanity` test needs a specific postgres version to test against because that's what our OIDs are from. If whoever wrote the oids and ran that function uses postgres 13, it'll differ if someone else happens to be running postgres 14.
- our current pgtest files pass against postgres 12. on postgres 13, cursors.pt fails due to statement tags changing. on postgres 14, copy-from-2.pt fails due to a connection close.

We should:
- decide on some postgres version (probably 14) and use it for all tests that can be affected by the version
- add an mzcompose workflow that spins up that postgres version
- it should test the pgtest directory
- unignore the verify_function_sanity function

**Comment by andrioni** [comment]
We currently declare ourselves to be Postgres 9.6, and some tools (like psql and sqlalchemy) actually change their behavior based on the self-reported version, so broad warning that we might also lose some compatibility if we decide to change that.

(we could, of course, continue claiming to be Postgres 9.6 and support APIs from newer versions, like we already do)

**Comment by benesch** [comment]
> the verify_function_sanity test needs a specific postgres version to test against because that's what our OIDs are from. If whoever wrote the oids and ran that function uses postgres 13, it'll differ if someone else happens to be running postgres 14.

So this isn't quite true, actually! OIDs less than 12000 are stable. They will never change across PostgreSQL versions. OIDs between 12000 and 16383 are allocated during database bootstrap. I think they can be different across different installations of PostgreSQL even if they have identical versions. We just need to be careful not to make any assertions about the OIDs in this range corresponding to any particular objects. That's what the test in #10684 does.

**Comment by benesch** [comment]
Huh, no, I guess I'm completely wrong. Function OIDs _do_ seem to move around between versions. Just not type OIDs.

**Comment by benesch** [comment]
Yeah, I think we're good to use v14 as the version we run pgtest/OID verification against. That's what I whipped up in #10684. Good to just always track the latest version there. If PostgreSQL is making visible changes to its interface in the protocol, error messages, OIDs, etc., we have license to make the identical changes to Materialize.

What PostgreSQL `server_version` we claim is an almost entirely separate question. That's dictated by the adapter tests (e.g., misc/dbt-materialize, the stuff in test/lang, etc.). I think generally we want to keep that as _low_ as possible, bumping it only when we need to e.g. get some adapter to recognize that we have support for some feature added in PostgreSQL 10.0. But a lot of adapters do proper feature detection rather than reasoning about what features exist based on the reported `server_version`, so it doesn't come up too much in practice.
